File: Cpp-Codes/DoublyLinkedList.cpp
// Implementing a Doubly Linked List with it's basic operations
// Whereever the 'change' is commented, means there is change else is same like SinglyLinkedList 
#include <iostream>
using namespace std;

// Node class
class Node{
    public:
        int val;
        Node *next;
        Node *prev;

        // Initilizing node
        Node(int v = 0){
            val = v;
            next = nullptr;
            prev = nullptr;
        }

        // Free, allocated memory
        ~Node(){}
};

// Linked list class
class Doubly{
    private:
        Node* head;

    public:

    Doubly() { head = nullptr; }    // initilize head
    // Destructor
    ~Doubly() {
        while(head != nullptr){
            Node* temp = head;
            head = head->next;
            delete temp;
        }
    }

    // List operations

    // display the list
    void display() const;
    // put val at last
    void push(int val);
    // remove value at last
    int pop();
    // insert val at specific index
    void insert(int index, int val);
    // remove a specific value
    void remove(int val);
    // get length of list
    int length() const;
    // get value at specific index
    int get(int index) const;
    // update value at specific index
    void update(int index, int val);
    // find min
    int mini();
    //find
    int maxi();
    // remove min val
    void rem_min();
    // find duplicate
    int duplicate();
    // reverse the list
    void rev_list();
};
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// List operations
    
// display the list
void Doubly::display() const{
    Node* temp = head;
    cout << "[ ";
    while(temp != nullptr){
        cout << temp->val << ", ";
        temp = temp->next;
    }
    cout << "]" << endl;
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// put val at last
void Doubly::push(int val){
    Node *new_node = new Node(val);

    // case 1, if no value current
    if(head == nullptr){
        head = new_node;
        return;
    }

    // case 2, if there is atleast one value
    Node* temp = head;
    while(temp->next != nullptr){
        temp = temp->next;
    }

    temp->next = new_node;
    new_node->prev = temp;
    return;
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// pop/tail value at last
int Doubly::pop(){
    // case 1: if no node
    if(head == nullptr){
        throw runtime_error("List is empty");
    }

    // case 2: if one node
    if(head->next == nullptr){
        Node* temp = head;
        int val = head->val;
        delete temp;
        head = nullptr;
        return val;
    }

    // case 3: traverse through last
    Node* prev = nullptr;         // to store the last state of temp
    Node* temp = head;
    while(temp->next != nullptr){
        prev = temp;
        temp = temp->next;
    }

    prev->next = nullptr;
    temp->prev = nullptr;
    int val = temp->val;
    delete temp;
    return val;
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// insert val at specific index
void Doubly::insert(int index, int val){
    Node* new_node = new Node(val);
   
    // case 1: if index is 0
    if(index == 0){
        new_node->next = head;

        if(head != nullptr){
            head->prev = new_node;
        }
        head = new_node;
        return;
    }

    // case 2: if index is in mid or end
    Node* prev = nullptr;
    Node* temp = head;
    int counter = 0;
    while(temp != nullptr && counter < index){
        prev = temp;
        temp = temp->next;
        counter += 1;
    }

    // if invalid index
    if(temp == nullptr){
        delete new_node;
        throw runtime_error("Invalid index!");
    }
    else{
        prev->next = new_node;
        new_node->prev = prev;

        new_node->next = temp;
        if(temp != nullptr){
            temp->prev = new_node;
        }
    } 
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// remove a specific value
void Doubly::remove(int val){

    Node* temp = head;

    // case 1: if no nodes
    if(head == nullptr){
        throw runtime_error("List is empty!");
    }

    // case 2: if sigle node
    if(temp->val == val){
        head = head->next;

        if(head->next != nullptr){
            head->prev = nullptr;
            delete temp;
        }
        return;
    }

    //case 3: if mid or end nodes
    Node* prev = nullptr;
    int counter = 0;
    while (temp != nullptr){
        if(temp->val == val){
            break;
        }

        prev = temp;
        temp = temp->next;
    }

    if(temp == nullptr){
        cout << "Value not found in the list!\n";
        return;
    }

    prev->next = temp->next;
    if(temp->next != nullptr){
        temp->next->prev = prev;
    }
    delete temp;
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// get length of list
int Doubly::length() const{
    Node* temp = head;

    // case 1: if empty
    if(head == nullptr){
        return 0;
    }

    // case 2: if list has some value
    int counter = 0;
    while (temp != nullptr){
        temp = temp->next;
        counter += 1;
    }
    return counter;
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// get value at specific index
int Doubly::get(int index) const{
    // case 1: if no nodes
    if(head == nullptr){
        throw runtime_error("List is empty!");
    }

    // case 2: if one value
    if(index == 0){
        return head->val;
    }

    // case 3: if more then one value
    Node* temp = head;
    int counter = 0;
    while(temp != nullptr && counter < index){
        temp = temp->next;
        counter += 1;
    }

    if(temp == nullptr){
        throw runtime_error("Index out of range.");
    }

    return temp->val;
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

// update value at specific index
void Doubly::update(int index, int val){
    // case 1: if no node
    if(head == nullptr){
        throw runtime_error("can't update, no value!");
    }

    // case 2: if index is 0
    if(index == 0){
        head->val = val;
        return;
    }

    // case 3: if any other index
    Node* temp = head;
    int counter = 0;
    while(temp != nullptr && counter < index){
        temp = temp->next;
        counter += 1;
    }

    temp->val = val;
    return;
}
// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX


int Doubly::mini(){
    // if list is empty
    if(head == nullptr){
        cout << "List is empty!" << endl;
        return -1;
    }

    // if single value
    if(head->next == nullptr){
        return head->val;
    }

    // for else
    int min_val = head->val;
    Node* temp = head->next;
    while(temp != nullptr){
        if(temp->val < min_val)
            min_val = temp->val;
        
        temp = temp->next;
    }

    return min_val;
}

// find max 
int Doubly::maxi(){
    // if list is empty
    if(head == nullptr){
        cout << "List is empty!" << endl;
        return -1;
    }

    // if single value
    if(head->next == nullptr){
        return head->val;
    }

    // for else
    int max_val = head->val;
    Node* temp = head->next;
    while(temp != nullptr){
        if(temp->val > max_val)
            max_val = temp->val;
        
        temp = temp->next;
    }

    return max_val;
}

// remove min val
void Doubly::rem_min(){
    remove(mini());
}

// find duplicate
int Doubly::duplicate(){
    // list is empty
    if(head == nullptr){
        cout << "List is empty!" << endl;
        return 0;
    }

    // if single value
    if(head->next == nullptr){
        cout << "List has only one value!" << endl;
        return head->val;
    }

    // else case
    Node* temp = head;
    while(temp != nullptr){
        int key = temp->val;

        Node* tempo = temp->next;
        while(tempo != nullptr){
            if(tempo->val == key)
                return key;
        
            tempo = tempo->next;
        }

        temp = temp->next;
    }
 
    // not found
    return -1;
}

// reverse the list
void Doubly::rev_list(){
    // if list is empty or single valued, list is reversed
    if(head == nullptr) return;
    if(head->next == nullptr) return;

    // if two or more vals
    // 2nd: get the last node & make it new_head
    Node* temp = head;
    while(temp->next != nullptr){
        temp = temp->next;
    }
    Node* new_head = temp;      // tail is the new_head 

    // reverse the list
    temp = head;
    while(temp != nullptr){
        // simple
        // swap(temp->next, temp->prev);
        // temp = temp->prev;      //after swap, old next is now stored in prev

        // mannually swapped 
        Node* hold = temp->next;
        temp->next = temp->prev;
        temp->prev = hold;

        temp = hold;     // hold has store the old state of temp->next     
    }

    head = new_head;            // change the head
}

int main(){
    Doubly l;

    // add vals to list
    l.push(2);
    l.push(4);
    l.push(7);
    l.push(5);
    l.push(4);
    l.display();

    l.rev_list();
    l.display();

    cout << l.duplicate() << endl;

    return 0;
}

File: Python-Codes/DoublyLinkedList.py
# Implementing a Doubly Linked List with it's basic operations
# Whereever the 'change' is commented, means there is change else is same like SinglyLinkedList 

# class for sigle node
class Node:
    def __init__(self, v = None):
        self.val = v
        self.next = None
        self.prev = None    # changing, we also want to secure address of previues node

# class for DoublyLinkedList
class Doubly:
    def __init__(self):
        self.head = None

# Operations, that will perform by DoublyLinkedList
# All the functions will be added to 'Doubly' class at run time

# Method to print List
def __str__(self):
    ret_str = '[ '
    temp = self.head
    while temp is not None:
        ret_str += str(temp.val) + ', '
        temp = temp.next

    ret_str = ret_str.rstrip(', ')
    ret_str += ' ]'
    return ret_str

Doubly.__str__ = __str__

# Methode to Push a value in list
# def push(self, val):
#     new_node = Node(val)

#     # case 1: if no Node before
#     if self.head is None:
#         self.head = new_node
#         return
    
#     # case 2: any other case
#     last = self.head
#     while last.next is not None:
#         last = last.next

#     last.next = new_node
#     new_node.prev = last        # change, we preserv last node also
#     return
# 
def push(self, val):
    self.insert(self.len(), val) 

Doubly.push = push

# Methode to pop a val from list
def pop(self):
    # case 1: if no node
    if self.head is None:
        raise Exception("can't pop, no value!")
    
    # case 2: if only one node exist
    if self.head.next is None:
        val = self.head.val
        self.head = None
        return val
    
    # case 3: if any other case
    temp = self.head
    while temp.next is not None:
        prev = temp
        temp = temp.next

    val = temp.val

    prev.next = None
    temp.prev = None        # changing

    return val

Doubly.pop = pop

# Insert a val in list 
def insert(self, index, val):
    new_node = Node(val)

    # case 1: if index is 0
    if index == 0:
        new_node.next = self.head
        
        if self.head is not None:           #change
            self.head.prev = new_node
    
        self.head = new_node
        return

    # case 2: other indecies
    counter = 0
    temp = self.head
    while temp is not None and counter < index:
        prev = temp
        temp = temp.next
        counter += 1
    
    prev.next = new_node
    new_node.prev = prev    # change


    new_node.next = temp
    if temp is not None:
        temp.prev = new_node    # change
    
Doubly.insert = insert

def remove(self, val):
    # case 1: if no node
    if self.head is None:
        raise Exception("can't remove, no value!")
    
    # case 2: if val find at first index
    if self.head.val == val:
        self.head = self.head.next
        
        if self.head.next is not None:
            self.head.prev = None

        return

    # case 3: if val at any other index
    temp = self.head
    while temp is not None:
        # if val find before list end
        if temp.val == val:
            break

        prev = temp
        temp = temp.next

    if temp is None:
        print("Value not found in the list!")
        return

    prev.next = temp.next
    if temp.next is not None:       # change
        temp.next.prev = prev
    
    return

Doubly.remove = remove

# length methd, to the list length
def len(self):
    temp = self.head

    #if no values
    if self.head is None:
        return 0
    
    #if list has values
    counter = 0
    while temp is not None:
        temp = temp.next
        counter += 1

    return counter

Doubly.len = len

# get a val at specific index
def get(self, index):
    # case 1: if no node
    if self.head is None:
        raise Exception("can't remove, no value!")
    
    # case 2: if index is 0
    if index == 0:
        return self.head.val
    
    # case 3: if any other index
    temp = self.head
    counter = 0
    while temp is not None and counter < index:
        temp = temp.next
        counter += 1

    return temp.val


Doubly.get = get

# to update a val at specific index
def update(self, index, val):
    # case 1: if no node
    if self.head is None:
        raise Exception("can't update, no value!")
    
    # case 2: if index is 0
    if index == 0:
        self.head.val = val
        return

    # case 3: if any other index
    temp = self.head
    counter = 0
    while temp is not None and counter < index:
        temp = temp.next
        counter += 1
    
    temp.val = val
    return


Doubly.update = update

# Testing List
if __name__ == "__main__":
    l = Doubly()

    # push a val
    l.push(2)
    l.push(3)
    print(l)

    # pop a val
    l.pop()
    l.pop()
    # l.pop()
    # l.remove(4)
    print(l)

    # insert a val
    l.insert(0, 80)
    l.insert(0, 90)
    l.insert(1, 100)
    l.insert(3, 68)
    l.push(80)
    print(l)

    # remove a val
    l.remove(90)
    l.remove(80)
    l.remove(80)
    # l.remove(8)
    print(l)

    # length methd, to the list length
    print(f"Length of list: {l.len()}")

    # get a val at specific index
    print(f"Value at index 0: {l.get(0)}")
    print(f"Value at index 0: {l.get(1)}")

    l.push(99)
    # update value
    print(l)
    l.update(0, 80)
    print(l)
    l.update(2, 80)
    print(l)